crates/ruma-api/src/lib.rs

```diff
@@ -562,3 +595,58 @@ impl MatrixVersion {
         }
     }
 }
+
+/// A specifier for what path version of an endpoint to request.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum EndpointPath {
+    /// Select an appropriate path version for a particular matrix version.
+    ForMatrixVersion(MatrixVersion),
+    /// Always select the stable path. (Will fail if the endpoint is still unstable)
+    PreferStable,
+}
+
+// This function needs to be public, yet hidden, as all `try_into_http_request`s would be using it.
+// We're also using MatrixVersion::repr() here, which is crate-private.
+#[doc(hidden)]
+pub fn select_path<'a>(
```

Comment:
I have some thoughts on changing this;

- Have `PreferStable` never emit an error, ever.
  - Have this try and return `stable` or `r0` or `unstable` in order.
- Add a `ForLatest` to the `EndpointPath`
  - Functionally the same to `PreferStable`, only it'll error if the endpoint has been removed.